### PR TITLE
[inductor] Preserve dtype in scatter_upon_const_tensor rewrite (#186481)

### PR DESCRIPTION
**Pull request by crystalwhale1024-meta**
Summary:

The `scatter_upon_const_tensor` joint-graph optimization (`caffe2/torch/_inductor/fx_passes/joint_graph.py`, gated by `config.optimize_scatter_upon_const_tensor`, on by default) rewrites `aten.full` + `aten.scatter.value` into a pointwise `torch.where(mask, val, background_val)`. When `val` and `background_val` are Python scalars, `torch.where` promotes the result to the default floating dtype (`float32`), which silently drops the const tensor's dtype. The `dtype` captured from the matched `aten.full` was a parameter of the rewrite function but was never used in `repl_fn`.

For a low-precision const tensor (f16/bf16) this changed the scatter result dtype from f16 to f32, which then promoted the whole dependent chain (and the op's output buffer) to f32. `aten.scatter.value` semantics require the result to have `self`'s dtype, with the scalar value cast to it, so this rewrite was not dtype-preserving.

This surfaced as a numerical failure of the `nll_loss_backward` opinfo on MTIA (f16): its decomposition emits `full_like(f16)` + `scatter.value(value=-1.0)` + `mul`, and the rewrite produced an f32 output buffer that the runtime read back as the op's declared f16, yielding garbage (each value landed at column `2*target`, the low/high 16 bits of the f32). The fix casts the `torch.where` result back to the captured `dtype`.

All existing tests in `test_scatter_optimization.py` used `dtype=torch.float`, so the where-with-f32-scalars coincidentally matched the f32 const tensor and the bug never surfaced; added `test_dtype_preserved` covering f16 and bf16.

Test Plan:
Unit test:
```
buck2 test fbcode//mode/opt fbcode//caffe2/test/inductor:test_scatter_optimization
```

Reviewed By: nandesuka

Differential Revision: D107587288

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by crystalwhale1024-meta** [comment]
@claude Is the test failure related to this PR ?

**Review by eellison — APPROVED** [review]

**Comment by crystalwhale1024-meta** [comment]
@claude are the failures related to this PR ?

**Comment by crystalwhale1024-meta** [comment]
@claude Are the failures in [trunk / linux-jammy-rocm-py3.10-mi355 / test (distributed, 1, 3, linux.rocm.gpu.gfx950.2)](https://github.com/pytorch/pytorch/actions/runs/27167114620/job/80222569860) related to this PR ?
